theatre/studio/src/uiComponents/usePresence.tsx

```diff
@@ -0,0 +1,231 @@
+            const rels = related && current && related[current]
+            if (rels) {
+              // can this be cached into a derived atom?
+              let best: HoveredWithImportance | undefined
+              for (const rel of Object.values(rels)) {
+                if (!rel) continue
+                if (best && best.importance >= rel.importance) continue
+                best = rel
+              }
+              return best?.relationship
+            }
+            return undefined
+          }
+        })
+      }, [itemKey])
+      return useVal(focusD)
+    },
+    setUserHover(itemKeyOpt) {
+      const prev = currentPresence.get()
+      if (prev === itemKeyOpt) {
+        return
+      }
+      if (prev) {
+        atom.setIn([prev], false)
+      }
+      currentPresence.set(itemKeyOpt)
+      if (itemKeyOpt) {
+        atom.setIn([itemKeyOpt], true)
+      }
+    },
+  }
+}
+
+type RelationshipConfig = {
+  affects: StudioSheetItemKey
+  relationship: Hovered
+  importance?: number
+}
+
+type Presence = {
+  useFocused(itemKey: StudioSheetItemKey | undefined): Hovered | undefined
+  setUserHover(itemKey: StudioSheetItemKey | undefined): void
+  /**  */
+  addRelationships(
+    itemKey: StudioSheetItemKey,
+    config: Array<RelationshipConfig>,
+  ): () => void
+}
+
+const presenceCtx = React.createContext<Presence>(createPresence())
+export function ProvidePresenceRoot({
+  presence = createPresence(),
+  children,
+}: React.PropsWithChildren<{presence?: Presence}>) {
+  return React.createElement(
+    presenceCtx.Provider,
+    {children, value: presence},
+    children,
+  )
+}
+
+const PRESENCE_ITEM_DATA_ATTR = 'data-pi'
+// const PRESENCE_ITEM_DEBUG_DATA_ATTR = 'data-pi-loc'
+
+export default function usePresence(config: {
+  key: StudioSheetItemKey | undefined
+}): [
+  props: {[attr: `data-${string}`]: string},
+  focus: {
+    current: Hovered | undefined
+    useRelationships(
+      getRelationships: () => Array<RelationshipConfig>,
+      // QUESTION: is this right?
+      deps: any[],
+    ): void
+  },
+] {
+  const presence = React.useContext(presenceCtx)
+  const focus = presence.useFocused(config.key)
+
+  return [
+    {
+      [PRESENCE_ITEM_DATA_ATTR]: config.key as string,
+      // [PRESENCE_ITEM_DEBUG_DATA_ATTR]: config.areaDebugName,
+    },
+    {
+      current: focus,
+      useRelationships(getRelationships, deps) {
+        useEffect(() => {
+          console.log('deps', deps.length, ...deps)
+          return (
+            config.key &&
+            presence.addRelationships(config.key, getRelationships())
+          )
+        }, [config.key, ...deps])
+      },
+    },
+  ]
+}
+
+export function usePresenceListeners(target: HTMLElement | null | undefined) {
+  const presence = React.useContext(presenceCtx)
+  const logger = useLogger('PresenceListeners')
+  useEffect(() => {
+    let cleanup: (() => void) | undefined = undefined
+    let last: any
```

Comment:
why are we keeping track of last? Please add comment

Reply:
Renamed to `currentItemKeyHovered` + comment "keep track of current hover to make sure we make changes to presence distinct"